### ISSUE TITLE
[基本面因子融合] | Fundamental Factor Fusion

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4467]
**标签**: quantitative-strategy
### 摘要
融合多维度基本面因子（估值、盈利、质量、动量）构建Alpha模型，使用IC加权动态调整因子权重。
### 算法描述
**核心原理**：
基本面因子可分解为估值、盈利、质量、成长四类。通过因子融合减少单一因子失效风险。

**五大类因子**：
1. **估值因子**：
   - PE, PB, PCF, PS
   - EV/EBITDA, Debt/Equity
2. **盈利因子**：
   - ROE, ROA, ROC
   - GrossMargin, NetMargin
3. **质量因子**：
   - AssetTurnover, CurrentRatio
   - Accruals, CFO/PnetIncome
4. **成长因子**：
   - RevenueGrowth, EPGGrowth
   - PEG, EarningsRevision
5. **动量因子**：
   - PriceMomentum, EarningsMomentum
   - RevenueRevision

**IC融合方法**：
IC = Corr(Factor, Future_Return)

**动态权重**：
W_i = IC_i / Σ|IC_j| * rolling_IC_std

**风控要点**：
- 市值中性化处理
- 行业偏差控制
- 因子共线性检测
### Python代码示例
```python
import numpy as np
import pandas as pd
from datetime import datetime, timedelta
from typing import Dict, List, Tuple, Optional
from dataclasses import dataclass
from enum import Enum
from sklearn.linear_model import LinearRegression
import statsmodels.api as sm

class FactorCategory(Enum):
    VALUE = "value"           # 估值
    PROFITABILITY = "profit"  # 盈利
    QUALITY = "quality"       # 质量
    GROWTH = "growth"         # 成长
    MOMENTUM = "momentum"     # 动量

@dataclass
class FactorData:
    code: str
    name: str
    timestamp: datetime
    # 估值因子
    pe: float
    pb: float
    pcashflow: float
    psales: float
    ev_ebitda: float
    debt_equity: float
    # 盈利因子
    roe: float
    roa: float
    gross_margin: float
    net_margin: float
    # 质量因子
    asset_turnover: float
    current_ratio: float
    accruals: float
    cfo_net_income: float
    # 成长因子
    revenue_growth: float
    earnings_growth: float
    peg: float
    # 动量因子
    price_momentum_3m: float
    price_momentum_12m: float
    earnings_revision: float

@dataclass
class FactorICResult:
    factor_name: str
    ic: float
    ic_ir: float  # IC的IR比率
    rank_ic: float
    rank_ic_ir: float
    cum_return: float

@dataclass
class PortfolioPosition:
    code: str
    weight: float
    factor_score: float
    entry_date: datetime

class FundamentalFactorModel:
    """基本面因子模型"""
    
    def __init__(
        self,
        lookback_ic: int = 60,
        rebalance_period: int = 20,
        min_weight: float = 0.01,
        max_weight: float = 0.05
    ):
        self.lookback_ic = lookback_ic
        self.rebalance_period = rebalance_period
        self.min_weight = min_weight
        self.max_weight = max_weight
        
        self.ic_history: Dict[str, List[float]] = {}
        self.factor_weights: Dict[str, float] = {}
        self.positions: List[PortfolioPosition] = []
    
    def calculate_value_factors(self, data: FactorData) -> Dict[str, float]:
        """计算估值因子"""
        return {
            "pe_inv": 1.0 / max(data.pe, 0.1) if data.pe > 0 else 0,
            "pb_inv": 1.0 / max(data.pb, 0.1) if data.pb > 0 else 0,
            "pcf_inv": 1.0 / max(data.pcashflow, 0.1) if data.pcashflow > 0 else 0,
            "ps_inv": 1.0 / max(data.psales, 0.1) if data.psales > 0 else 0,
            "ev_ebitda_inv": 1.0 / max(data.ev_ebitda, 0.1) if data.ev_ebitda > 0 else 0,
            "debt_equity_inv": 1.0 / max(data.debt_equity, 0.1) if data.debt_equity > 0 else 0
        }
    
    def calculate_profitability_factors(self, data: FactorData) -> Dict[str, float]:
        """计算盈利因子"""
        return {
            "roe": data.roe,
            "roa": data.roa,
            "gross_margin": data.gross_margin,
            "net_margin": data.net_margin
        }
    
    def calculate_quality_factors(self, data: FactorData) -> Dict[str, float]:
        """计算质量因子"""
        return {
            "asset_turnover": data.asset_turnover,
            "current_ratio": data.current_ratio,
            "accruals": -data.accruals,  # 低应计项目为正
            "cfo_net_income": data.cfo_net_income
        }
    
    def calculate_growth_factors(self, data: FactorData) -> Dict[str, float]:
        """计算成长因子"""
        return {
            "revenue_growth": data.revenue_growth,
            "earnings_growth": data.earnings_growth,
            "peg_inv": data.peg if data.peg < 0 else 1.0 / max(data.peg, 0.1)
        }
    
    def calculate_momentum_factors(self, data: FactorData) -> Dict[str, float]:
        """计算动量因子"""
        return {
            "price_momentum_3m": data.price_momentum_3m,
            "price_momentum_12m": data.price_momentum_12m,
            "earnings_revision": data.earnings_revision
        }
    
    def normalize_factor(self, values: np.ndarray) -> np.ndarray:
        """标准化因子值（Z-score）"""
        if len(values) < 2:
            return np.zeros_like(values)
        
        mean = np.mean(values)
        std = np.std(values)
        
        if std < 1e-10:
            return np.zeros_like(values)
        
        return (values - mean) / std
    
    def calculate_information_coefficient(
        self,
        factor_values: np.ndarray,
        future_returns: np.ndarray
    ) -> float:
        """计算IC（信息系数）"""
        if len(factor_values) != len(future_returns) or len(factor_values) < 10:
            return 0.0
        
        # Spearman秩相关
        return np.corrcoef(
            np.argsort(np.argsort(factor_values)),
            np.argsort(np.argsort(future_returns))
        )[0, 1]
    
    def calculate_factor_ic_series(
        self,
        factor_data: pd.DataFrame,
        return_period: int = 20
    ) -> pd.DataFrame:
        """计算因子IC序列"""
        factor_cols = [col for col in factor_data.columns if col not in ["code", "date", "return"]]
        
        ic_results = {}
        
        for factor in factor_cols:
            ic_series = []
            
            for i in range(self.lookback_ic, len(factor_data)):
                window = factor_data.iloc[i-self.lookback_ic:i]
                
                factor_values = window[factor].values
                future_returns = window["return"].values
                
                ic = self.calculate_information_coefficient(factor_values, future_returns)
                ic_series.append(ic)
            
            ic_results[factor] = ic_series
        
        return pd.DataFrame(ic_results)
    
    def calculate_factor_weights(
        self,
        ic_results: pd.DataFrame
    ) -> Dict[str, float]:
        """计算因子权重"""
        weights = {}
        
        for factor in ic_results.columns:
            ic_values = ic_results[factor].dropna()
            
            if len(ic_values) < 10:
                weights[factor] = 1.0
                continue
            
            # IC均值作为权重
            ic_mean = np.mean(ic_values)
            ic_std = np.std(ic_values)
            
            # IR比率
            ir = ic_mean / ic_std if ic_std > 0 else 0
            
            # 权重 = IC均值 * IR
            weight = ic_mean * abs(ir)
            weights[factor] = max(0.01, weight)
        
        # 归一化
        total = sum(weights.values())
        if total > 0:
            for k in weights:
                weights[k] /= total
        
        return weights
    
    def calculate_composite_score(
        self,
        data: FactorData,
        weights: Dict[str, float],
        category_weights: Dict[FactorCategory, float]
    ) -> float:
        """计算综合因子得分"""
        all_factors = {}
        
        # 各类别因子
        value_factors = self.calculate_value_factors(data)
        profit_factors = self.calculate_profitability_factors(data)
        quality_factors = self.calculate_quality_factors(data)
        growth_factors = self.calculate_growth_factors(data)
        momentum_factors = self.calculate_momentum_factors(data)
        
        all_factors.update(value_factors)
        all_factors.update(profit_factors)
        all_factors.update(quality_factors)
        all_factors.update(growth_factors)
        all_factors.update(momentum_factors)
        
        # 加权求和
        score = 0.0
        for factor_name, factor_value in all_factors.items():
            w = weights.get(factor_name, 1.0)
            
            # 判断因子类别
            if factor_name in value_factors:
                cat_w = category_weights[FactorCategory.VALUE]
            elif factor_name in profit_factors:
                cat_w = category_weights[FactorCategory.PROFITABILITY]
            elif factor_name in quality_factors:
                cat_w = category_weights[FactorCategory.QUALITY]
            elif factor_name in growth_factors:
                cat_w = category_weights[FactorCategory.GROWTH]
            else:
                cat_w = category_weights[FactorCategory.MOMENTUM]
            
            score += factor_value * w * cat_w
        
        return score
    
    def market_neutralize(
        self,
        factor_values: np.ndarray,
        market_caps: np.ndarray
    ) -> np.ndarray:
        """市值中性化"""
        if len(factor_values) != len(market_caps):
            return factor_values
        
        log_mcap = np.log(market_caps + 1)
        
        # 回归残差
        X = sm.add_constant(log_mcap)
        y = factor_values
        
        try:
            model = sm.OLS(y, X).fit()
            residuals = model.resid
            return residuals
        except:
            return factor_values
    
    def rank_and_weight(
        self,
        scores: List[Tuple[str, float]],
        method: str = "equal_weight"
    ) -> Dict[str, float]:
        """排名并分配权重"""
        # 按得分排序
        sorted_scores = sorted(scores, key=lambda x: x[1], reverse=True)
        
        n = len(sorted_scores)
        weights = {}
        
        for rank, (code, score) in enumerate(sorted_scores):
            if method == "equal_weight":
                weight = 1.0 / n
            elif method == "rank_weight":
                weight = (n - rank) / (n * (n + 1) / 2)
            else:
                weight = 1.0 / n
            
            # 应用权重限制
            weight = max(self.min_weight, min(self.max_weight, weight))
            weights[code] = weight
        
        return weights

class FactorFusionStrategy:
    """基本面因子融合策略"""
    
    def __init__(self, initial_capital: float = 10000000):
        self.capital = initial_capital
        self.model = FundamentalFactorModel()
        self.history = []
    
    def run_backtest(
        self,
        factor_data: pd.DataFrame,
        price_data: pd.DataFrame,
        rebalance_freq: int = 20
    ) -> Dict:
        """回测"""
        positions = []
        
        for i in range(60, len(factor_data), rebalance_freq):
            window = factor_data.iloc[max(0, i-60):i]
            
            # 计算IC和权重
            ic_results = self.model.calculate_factor_ic_series(window)
            factor_weights = self.model.calculate_factor_weights(ic_results)
            
            # 计算综合得分
            category_weights = {
                FactorCategory.VALUE: 0.25,
                FactorCategory.PROFITABILITY: 0.25,
                FactorCategory.QUALITY: 0.20,
                FactorCategory.GROWTH: 0.15,
                FactorCategory.MOMENTUM: 0.15
            }
            
            scores = []
            for _, row in window.iterrows():
                data = FactorData(
                    code=row["code"],
                    name=row.get("name", ""),
                    timestamp=row["date"],
                    pe=row.get("pe", 10),
                    pb=row.get("pb", 1),
                    pcashflow=row.get("pcf", 10),
                    psales=row.get("ps", 1),
                    ev_ebitda=row.get("ev_ebitda", 10),
                    debt_equity=row.get("debt_equity", 0.5),
                    roe=row.get("roe", 0.1),
                    roa=row.get("roa", 0.05),
                    gross_margin=row.get("gross_margin", 0.3),
                    net_margin=row.get("net_margin", 0.1),
                    asset_turnover=row.get("asset_turnover", 1),
                    current_ratio=row.get("current_ratio", 1.5),
                    accruals=row.get("accruals", 0),
                    cfo_net_income=row.get("cfo_net_income", 1),
                    revenue_growth=row.get("revenue_growth", 0.1),
                    earnings_growth=row.get("earnings_growth", 0.1),
                    peg=row.get("peg", 1),
                    price_momentum_3m=row.get("momentum_3m", 0),
                    price_momentum_12m=row.get("momentum_12m", 0),
                    earnings_revision=row.get("earnings_revision", 0)
                )
                
                score = self.model.calculate_composite_score(
                    data, factor_weights, category_weights
                )
                scores.append((row["code"], score))
            
            # 分配权重
            weights = self.model.rank_and_weight(scores)
            
            # 计算收益
            next_prices = price_data[price_data["date"] == factor_data.iloc[i]["date"]]
            
            self.history.append({
                "date": factor_data.iloc[i]["date"],
                "positions": len(weights),
                "weights": weights
            })
        
        return {
            "final_capital": self.capital,
            "total_return": (self.capital - 10000000) / 10000000,
            "history": self.history
        }

# 测试
if __name__ == "__main__":
    np.random.seed(42)
    
    codes = [f"00{i:04d}.SZ" for i in range(100)]
    dates = pd.date_range("2024-01-01", "2025-01-01", freq="M")
    
    data_rows = []
    
    for date in dates:
        for code in codes:
            base_return = np.random.randn() * 0.05
            
            data_rows.append({
                "date": date,
                "code": code,
                "name": f"Stock_{code}",
                "pe": np.random.uniform(5, 50),
                "pb": np.random.uniform(0.5, 5),
                "pcf": np.random.uniform(5, 30),
                "ps": np.random.uniform(0.5, 10),
                "ev_ebitda": np.random.uniform(5, 20),
                "debt_equity": np.random.uniform(0.1, 2),
                "roe": np.random.uniform(-0.2, 0.3),
                "roa": np.random.uniform(-0.1, 0.15),
                "gross_margin": np.random.uniform(0.1, 0.5),
                "net_margin": np.random.uniform(0.01, 0.2),
                "asset_turnover": np.random.uniform(0.5, 2),
                "current_ratio": np.random.uniform(0.5, 3),
                "accruals": np.random.uniform(-0.1, 0.1),
                "cfo_net_income": np.random.uniform(0.5, 1.5),
                "revenue_growth": np.random.uniform(-0.2, 0.5),
                "earnings_growth": np.random.uniform(-0.3, 0.6),
                "peg": np.random.uniform(0.5, 3),
                "momentum_3m": np.random.uniform(-0.15, 0.2),
                "momentum_12m": np.random.uniform(-0.4, 0.6),
                "earnings_revision": np.random.uniform(-0.1, 0.1),
                "return": base_return,
                "mcap": np.random.uniform(1e8, 1e10)
            })
    
    df = pd.DataFrame(data_rows)
    price_df = df[["date", "code", "return"]].copy()
    price_df["close"] = 10 * (1 + price_df["return"].cumsum())
    
    strategy = FactorFusionStrategy()
    result = strategy.run_backtest(df, price_df)
    
    print(f"Final Capital: ¥{result['final_capital']:,.2f}")
    print(f"Total Return: {result['total_return']:.2%}")
